Rates endpoint - TimeUnit interval (#294)
MewsSystems/gitbook-connector-api: operations/services.md

| `EndOffset` | string | required | Offset from the end of the time unit defining the default end of the service orders in ISO 8601 duration format. |
| `OccupancyStartOffset` | string | required | Offset from the start of the time unit defining the occupancy start of the service in ISO 8601 duration format that is considered regarding the availability and reporting. |
| `OccupancyEndOffset` | string | required | Offset from the end of the time unit defining the occupancy end of the service in ISO 8601 duration format that is considered regarding the availability and reporting. |
| `TimeUnit` | [Time unit period](#time-unit-period) | required | Time unit period of the service. |

> #### Time unit start and end

Reviewer: I do not understand why this paragraph is here. I have no context of time unit in this place. It should be under #Time unit definition

> Time units represent a fixed, finite time interval: a minute, a day, a month, etc. A Time unit defines the operable periods for a bookable service.

Reviewer: space under the hadding

> Time units represent a fixed, finite time interval: a minute, a day, a month, etc. A Time unit defines the operable periods for a bookable service.

Reviewer: This seems to be a duplicity with definition of TimeUnit
There is 
Time unit
* Day,
* Month 
The description should be there no?

| `OccupancyEndOffset` | string | required | Offset from the end of the time unit defining the occupancy end of the service in ISO 8601 duration format that is considered regarding the availability and reporting. |
| `TimeUnit` | [Time unit period](#time-unit-period) | required | Time unit period of the service. |

#### Time unit start and end
Time units represent a fixed, finite time interval: a minute, a day, a month, etc. A Time unit defines the operable periods for a bookable service.

> We think of the daily time unit as the physical time unit that starts at midnight and ends at midnight the following day. A monthly time unit is a time unit that starts at midnight on the first day of the month and ends at midnight on the first day of the following month.

Reviewer 2: > We think of the daily time unit as the physical time unit that starts at midnight and ends at midnight the following day.

I'm wondering if there's a different way to describe this? I had understood that midnight the following day means the start of the next day, so if I only wanted to indicate one daily time unit via the API, I should send the same Start/End time, or End time any time before midnight the next day. As a partner reading the documentation I might not know this and assume that sending midnight today to midnight the next day would indicate one daily time unit.

Reviewer 3: It should not longer be the case with the new API that drops StartUtc and EndUtc and work with FirstTimeUniT and LastTimeUnit no?
This paragraph is about describing what is the time unit. Then there is the picture that is aligned with what is described here.
From partner we then want to send StartUtc of the FirstTimeUnit they would like to affect and StartUtc of the LastTimeUnit. Do you think it is not understandable that for updating the one day the FirstTimeUnit and LastTimeUnit would be the same time unit? (with the same StartUtc).

Negative or zero end offsets of the daily time unit define the daily service as depicted on the picture below.
![](../.gitbook/assets/timeunits-connector-day.png)

#### Time unit period

> * `Day`

Reviewer 4: How about adding `Hour` to make it complete?

Author: We do not want to mention hour just yet, as there are still some open questions and we do not want to confuse partners.

Reviewer 5: We recently had the discussion about enums with ellipsis or without and about the implications. 
This definitely feels like the hard requirement which should have all values listed. But on the other hand I understand the situation with rollout. So let's keep it this way for now.

Reviewer 3: We could of course add it, but there will be no such service even for testing purposes probably. So then I am not 100% sure if it is ok to list it while it is not really supported. The thing with the monthly time unit is that it is not 100% supported either, but the pricing is in its end state so the partner can update those 2 endpoints and it will not change for the monthly unit. For hours, it is still not decided if you are going to update hour prices for every hour for every day. Maybe you will be updating just specific hours generally for all days the same or in some different batches, etc... That is the reason why we do not want to add it right now, even just for the pricing, because we cannot guarantee it will work in the same way at the moment,

Negative or zero end offsets of the daily time unit define the daily service as depicted on the picture below.
![](../.gitbook/assets/timeunits-connector-day.png)

#### Time unit interval length restrictions

> Interval length created between FirstTimeUnit StartUtc and LastTimeUnit StartUtc is allowed for up to 100 time units, but no more than 2 years.

Reviewer: space

## Get rate pricing

> Returns prices of a rate in the specified interval. Note that response contains prices for all dates that the specified interval intersects. So e.g. interval `1st Jan 13:00 - 1st Jan 14:00` will result in one price for `1st Jan`. Interval `1st Jan 23:00 - 2nd Jan 01:00` will result in two prices for `1st Jan` and `2nd Jan`. Interval length is allowed for up to 100 time units, but no more than 2 years.

Reviewer 3: This should be redefined as well. There is no longer possible to send different times than midnights, so the whole paragraph describing the dropping of hours does not make sense now.

## Get rate pricing

> Returns prices of a rate in the specified interval. Note that response contains prices for all dates that the specified interval intersects. So e.g. interval `1st Jan 00:00 - 1st Jan 00:00` will result in one price for `1st Jan`. Interval `1st Jan 00:00 - 2nd Jan 00:00` will result in two prices for `1st Jan` and `2nd Jan`. Other time part of interval than [time unit start](#time-unit-start-and-end) is not supported. Time unit max length [restrictions](#time-unit-interval-length-restrictions) applies.

Reviewer 5: ```suggestion
Returns prices of a rate in the specified interval. Note that response contains prices for all time units that the specified interval intersects. So e.g. interval `1st Jan 00:00 - 1st Jan 00:00` will result in one price for `1st Jan`. Interval `1st Jan 00:00 - 2nd Jan 00:00` will result in two prices for `1st Jan` and `2nd Jan`. Other time part of interval than [time unit start](#time-unit-start-and-end) is not supported. Time unit max length [restrictions](#time-unit-interval-length-restrictions) applies.
```

| `AccessToken` | string | required | Access token of the client application. |
| `Client` | string | required | Name and version of the client application. |
| `RateId` | string | required | Unique identifier of the [Rate](services.md#rate) whose prices should be returned. |

> | `FirstTimeUnit` | [TimeUnit](#time-unit) | required | First TimeUnit of the interval. |

Reviewer 3: There as max length 3 months mentioned here. Do not we want to keep mentioning it somehow? Maybe like required, max length - where this max length would be the link to some general max length paragraph describing the limitation of 100 units or 2 years? cc @vitasamek

| `AccessToken` | string | required | Access token of the client application. |
| `Client` | string | required | Name and version of the client application. |
| `RateId` | string | required | Unique identifier of the [Rate](services.md#rate) whose prices should be returned. |
| `FirstTimeUnit` | [TimeUnit](#time-unit) | required, [max length](#time-unit-interval-length-restrictions) | First TimeUnit of the interval. |

> | `LastTimeUnit` | [TimeUnit](#time-unit) | required, [max length](#time-unit-interval-length-restrictions) | Last TimeUnit of the interval. |

Reviewer 5: Similar here, the current contract wasn't described in a best way.

Author: Added the info about that to paragraph describing the max length.

            "Prices": [ 20, 20, 20 ]
        }
    ],
    "TimeUnitStartsUtc": [
        "2022-01-01T23:00:00Z"
        "2022-01-02T23:00:00Z"
        "2022-01-03T23:00:00Z"
    ]
}
```

| Property | Type | Contract | Description |
| --- | --- | --- | --- |
| `Currency` | string | required | ISO-4217 code of the [Currency](configuration.md#currency). |

> | `TimeUnitStartsUtc` | array of string | required | Covered time units starts in UTC timezone. |

Reviewer 3: Covered time unit starts in the UTC timezone in ISO 8601 format.

## Update rate price

> Updates price of a rate in the specified intervals. If the `CategoryId` is specified, updates price of the corresponding [Resource category](enterprises.md#resource-category), otherwise updates the base price for all resource categories. Note that prices are defined daily, so when the server receives the UTC interval, it first converts it to enterprise timezone and updates the price on all dates that the interval intersects. Only root rates can be updated (the rates that have no base rate, that have `BaseRateId` set to `null`). It's not allowed to update past prices outside of `EditableHistoryInterval`. Time unit max length [restrictions](#time-unit-interval-length-restrictions) applies.

Reviewer 5: ```suggestion
Updates price of a rate in the specified intervals. If the `CategoryId` is specified, updates price of the corresponding [Resource category](enterprises.md#resource-category), otherwise updates the base price for all resource categories. Note that prices are defined per time unit (i.e. daily), so when the server receives the UTC interval, it first converts it to enterprise timezone and updates the price on all time units that the interval intersects. Only root rates can be updated (the rates that have no base rate, that have `BaseRateId` set to `null`). It's not allowed to update past prices outside of `EditableHistoryInterval`. Time unit max length [restrictions](#time-unit-interval-length-restrictions) applies.
```

| Property | Type | Contract | Description |
| --- | --- | --- | --- |
| `CategoryId` | string | optional | Unique identifier of the [Resource category](enterprises.md#resource-category) whose prices to update. If not specified, base price is updated. |

> | `FirstTimeUnit` | [TimeUnit](services.md#time-unit) | required | First TimeUnit of the interval. |
> | `LastTimeUnit` | [TimeUnit](services.md#time-unit) | required | Last TimeUnit of the interval. |

Reviewer 4: ```suggestion
| `FirstTimeUnit` | [TimeUnit](#time-unit) | required | First TimeUnit of the interval. |
| `LastTimeUnit` | [TimeUnit](#time-unit) | required | Last TimeUnit of the interval. |
```
No need to specify same .md file

| Property | Type | Contract | Description |
| --- | --- | --- | --- |
| `CategoryId` | string | optional | Unique identifier of the [Resource category](enterprises.md#resource-category) whose prices to update. If not specified, base price is updated. |
| `FirstTimeUnit` | [TimeUnit](#time-unit) | required, [max length](#time-unit-interval-length-restrictions) | First TimeUnit of the interval. |

> | `LastTimeUnit` | [TimeUnit](#time-unit) | required, [max length](#time-unit-interval-length-restrictions) | Last TimeUnit of the interval. |

Reviewer 5: This isn't technically correct as `TimeUnit` object doesn't have interval length. Maybe just mention that the difference between `FirstTimeUnit` and `LastTimeUnit` is the max lenght?

Author: Added the info about that to paragraph describing the max length.
